dynamiqs/utils/utils.py

```diff
@@ -220,27 +259,36 @@ def tensprod(*args: Tensor) -> Tensor:
     The returned tensor shape is:
 
     - $(n, 1)$ with $n=\prod_k n_k$ if all input tensors are kets with shape $(n_k, 1)$,
+    - $(1, n)$ with $n=\prod_k n_k$ if all input tensors are bras with shape $(1, n_k)$,
     - $(n, n)$ with $n=\prod_k n_k$ if all input tensors are density matrices or
       operators vectors with shape $(n_k, n_k)$.
 
-    Warning:
-        This function does not yet support tensors in the bra format.
-
-    Warning:
-        This function does not yet support arbitrarily batched tensors (see
-        [issue #69](https://github.com/dynamiqs/dynamiqs/issues/69)).
-
     Note:
         This function is the equivalent of `qutip.tensor()`.
 
     Args:
         *args (n_k, 1) or (n_k, n_k): Sequence of kets, density matrices or operators.
 
     Returns:
-        (n, 1) or (n, n): Tensor product of the input tensors.
+        (n, 1) or (1, n) or (n, n): Tensor product of the input tensors.
     """
-    # TODO: adapt to bras
-    return reduce(torch.kron, args)
+    return reduce(_bkron, args)
+
+
+def _bkron(x: Tensor, y: Tensor) -> Tensor:
+    """Compute the batched Kronecker product of two matrices."""
+    x_type = _quantum_type(x)
+    y_type = _quantum_type(y)
+    if x_type != y_type:
+        raise TypeError(
+            'Cannot take the tensor product of different input objects. `x` is a'
+            f' {x_type} and `y` is a {y_type}.'
+        )
+
+    kron_dims = torch.Size(torch.tensor(x.shape[-2:]) * torch.tensor(y.shape[-2:]))
+    out = x.unsqueeze(-1).unsqueeze(-3) * y.unsqueeze(-2).unsqueeze(-4)
+    batch_dims = out.shape[:-4]
+    return out.reshape(batch_dims + kron_dims)
```

Comment:
Maybe we could add some comments because this code is a bit cryptic:
```suggestion
    # x: (..., x1, x2)
    # y: (..., y1, y2)
    
    batch_dims = x.shape[:-2]
    x1, x2 = x.shape[-2:]
    y1, y2 = y.shape[-2:]
    kron_dims = torch.Size((x1 * y1, x2 * y2))

    # perform element-wise multiplication of appropriately unsqueezed tensors to
    # simulate the Kronecker product 
    x_tmp = x.unsqueeze(-1).unsqueeze(-3)  # (..., x1, 1, x2, 1)
    y_tmp = y.unsqueeze(-2).unsqueeze(-4)  # (..., 1, y1, 1, y2)
    out = x_tmp * y_tmp  # (..., x1, y1, x2, y2)

    # reshape the output
    return out.reshape(batch_dims + kron_dims) # (..., x1 * y1, x2 * y2)
```